Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/CruiseControl.cs

```diff
@@ -0,0 +1,2277 @@
+            SpeedRegMode = (SpeedRegulatorMode)fSpeedRegMode;
+            int fSpeedSelMode = inf.ReadInt32();
+            SpeedSelMode = (SpeedSelectorMode)fSpeedSelMode;
+            throttleIsZero = inf.ReadBoolean();
+            trainBrakePercent = inf.ReadSingle();
+            TrainLengthMeters = inf.ReadInt32();
+            speedRegulatorIntermediateValue = inf.ReadSingle();
+            CCIsUsingTrainBrake = inf.ReadBoolean();
+        }
+
+        public void UpdateSelectedSpeed(float elapsedClockSeconds)
+        {
+            TotalTime += elapsedClockSeconds;
+            if (SpeedRegMode == CruiseControl.SpeedRegulatorMode.Auto && !DynamicBrakePriority ||
+             EnableSelectedSpeedSelectionWhenManualModeSet)
+            {
+                if (selectedSpeedIncreasing) SpeedRegulatorSelectedSpeedIncrease();
+                if (SelectedSpeedDecreasing) SpeedRegulatorSelectedSpeedDecrease();
+            }
+        }
+
+
+        public void SpeedRegulatorModeIncrease()
+        {
+            if (!Locomotive.IsPlayerTrain) return;
+            Locomotive.SignalEvent(Common.Event.CruiseControlSpeedRegulator);
+            SpeedRegulatorMode previousMode = SpeedRegMode;
+            if (!Equipped) return;
+            if (SpeedRegMode == SpeedRegulatorMode.Testing) return;
+            if (SpeedRegMode == SpeedRegulatorMode.Manual &&
+               ((DisableManualSwitchToAutoWhenThrottleNotAtZero && (Locomotive.ThrottlePercent != 0 || 
+               (Locomotive.DynamicBrakePercent != -1 && Locomotive.DynamicBrakePercent != 0))) ||
+               (DisableManualSwitchToAutoWhenSetSpeedNotAtTop && SelectedSpeedMpS != Locomotive.MaxSpeedMpS && Locomotive.AbsSpeedMpS > Simulator.MaxStoppedMpS)))
+                return;
+            bool test = false;
+            while (!test)
+            {
+                SpeedRegMode++;
+                switch (SpeedRegMode)
+                {
+                    case SpeedRegulatorMode.Auto:
+                        {
+                            if (SpeedRegulatorOptions.Contains("regulatorauto")) test = true;
+                            if (!DisableManualSwitchToAutoWhenSetSpeedNotAtTop && !KeepSelectedSpeedWhenManualModeSet) SelectedSpeedMpS = Locomotive.AbsSpeedMpS;
+                            break;
+                        }
+                    case SpeedRegulatorMode.Testing: if (SpeedRegulatorOptions.Contains("regulatortest")) test = true; break;
+                }
+                if (!test && SpeedRegMode == SpeedRegulatorMode.Testing) // if we're here, then it means no higher option, return to previous state and get out
+                {
+                    SpeedRegMode = previousMode;
+                    return;
+                }
+            }
+            Simulator.Confirmer.Message(ConfirmLevel.Information, Simulator.Catalog.GetString("Speed regulator mode changed to") + " " + Simulator.Catalog.GetString(SpeedRegMode.ToString()));
```

Comment:
It's better to use "Speed regulator mode changed to {0}" instead of string concatenation to allow localization for languages with different syntax.

Reply:
Done